### ISSUE TITLE
Looking to collaborate, have some questions

### ISSUE THREAD

**Issue by mcheah** [LoialOtter/Open-Lerdge-X#1]
Hey, thanks for posting the pinout, this saves me a whole bunch of work.
I developed an [STM32-based Marlin port for the Monoprice Mini Delta](https://github.com/mcheah/Marlin4MPMD) and just got a Lerdge board as it's much higher performance.  Porting should be pretty easy to do, but I have some questions about the system that I'm hoping you can answer.  Specifically I'd like to know how you bypassed the write protect without clearing the flash rom, and how you were able to probe the GPIO pin configuration registers.
Hit me up on [reddit](https://www.reddit.com/user/m_rozay) it'd be nice to work with somebody else on this.

**Comment by LoialOtter** [comment]
Very welcome. The board is really quite impressive for being so cheap.

The readout protection on the STM32 line is pretty good (I've developed products that made use of it) but there are some interesting issues with it too. One is that protection of RAM and peripherals is optional. I just and hooked up to JTAG and read from those locations on RAM. Note that it's very important that you do it carefully so you don't accidentally read from any location in flash as that will lock down the JTAG access and you'll no longer be able to read from anywhere. I was really not expecting that to work so I had also pulled out my multimeter to do things the old fashioned way.

Once I had the map done there's a way to erase the entire flash by disabling the readout protection. This then leaves you with a fully unlocked but fully erased device.

**Comment by mcheah** [comment]
Yeah, i did something similar on the stm32F0 firmware on the monoprice printer.  I see that both the bootloader and the application bin files are available, so I'm tempted to just blow it all away and start from scratch, but I want to make sure i have the memory map right first.  I'm assuming the bootloader ends at 64k, do you know anything about that?  Also do you have any info on the LCD / touch screen they used?  Tbh I'll likely never implement since the goal is to run off of octoprint.  Also do you know why they have so many frikkin mosfets all over the board?  Even for the stepper drivers.  I'd  understand for the heater/fan but it just seems excessive for digital inputs.  Maybe they need to level shift to 5V?

**Comment by LoialOtter** [comment]
The mosfets are proper driving and separation so any current spikes will never hit the CPU. They are proper mosfet drivers so also do level translation IIRC. Honestly the IO is probably the best designed I've ever seen in a reprap community driver board. Everything's been given proper inputs and outputs.

My only real complaint from a 3d printing side is that the heated bed driver isn't on the board so you'll need to add a SSR or mosfet or something for that.

The LCD is being driven by one of the DMAs rather than any special peripheral. This means they have a framebuffer in ram that they're simply drawing to and the DMA just copies that onto the display. The pinout should be pretty obvious. It's just a standard 8080/6800 LCD and there's specs online for the resistive touchpad chip on there; I didn't get too deep into it as I was going to go deal with that later and then got distracted from the project.

As for the bootloader and flash layout, no idea as I'm replacing the whole thing. Don't forget the CPU itself has DFU so you don't need anything special, just pull a pin low when you power it on and it'll come up in windows/linux and you can program - no special bootloader needed.

**Comment by mcheah** [comment]
I figured they were kind of superfluous since most stepper drivers support CMOS inputs, but I suppose in the case of a short or failure with the high voltage supplies going to them you'd want to protect the CPU.  Just a lot higher component count than I'd expect from a board this size and cost.  Agreed on the bed driver.  My suspicion is that they realized that it'd be difficult route that much power over a small 2-pin connector and opted to have it supplied externally.  While that's definitely the more flexible route, it's frustrating because the provided solution is a dongle to a dongle to supply a heatbed and potentially requires a second power supply or a jury rigged connection.  

Thanks for the info on the LCD.  I have no idea how to look up info on these screens, most the time they come with very few markings and a gigantic parallel bus.  It'd be nice if somebody made a nice library for these kind of things.  IMO the graphics and UI design for a product like this is probably 3x the work of the motor control part.  I'll likely ditch the whole thing and hook up a simple reprap display through the expansion connector.

Yeah, I've got a ST-Link hooked up through the SWD pins already, I need the ability to debug code directly so I'm skipping the DFU route.  What are your plans for the firmware?  Are you looking to adapt Marlin/RepRap/Klipper/Smoothie to Lerdge?  Or build your own from scratch?  
The other thing I didn't like about the board is that there's no native USB port for communication, you have a buy an adapter with a CH401 chip.  I tried plugging in a USB A<->A M/M but nothing showed up, I think they have it hooked up as host only.

I'd like to retain the ability to run the stock firmware so I can go back and forth to compare functionality.  This is how I was able to debug a 5 degree offset on the temperature readings on my Marlin project.

**Comment by LoialOtter** [comment]
https://github.com/LoialOtter/micropython

**Comment by LoialOtter** [comment]
M/M works fine. Not standard but hey, you just have to configure the system to not use OTG and instead force device (so a UI is needed if you want it to do host).

The DFU is there as a bootloader for users. I haven't tried erasing a chip using it though that would be interesting. It allows you to ignore their bootloader and just use that if you wish or make one yourself. I've made a few in the past based on ChibiOS which works great on this line of chips.

**Comment by mcheah** [comment]
Yeah, I'll get it working with M/M, but it's not ideal.  I don't mind losing the USB key, I don't quite see the need with an available SD slot.
DFU you still need to hook up to a PC and run some STM tool for flashing right?  It just runs some hard-coded bootloader in ROM if I recall, I don't think you can put in a custom routine to load off a USB flash disk and copy over firmware.bin for example.  So in that scenario you'd require the user to buy a USB M/M cable just for flashing, and I'd just as re-use their own bootloader at that point.

**Comment by LoialOtter** [comment]
DFU under linux is just a normal package - it's not just ST as well as DFU is a a standard for uploading firmware to USB devices.

For me USB is pretty important as it gives the command prompt for python. My plan is to make a lowlevel driver set which will get the steppers and temperature control reliably running then use python to do the motion planning and g-code to control interface. I want to go this route mostly due to wanting to have a platform that's super easy to get working with unusual hardware. As an added benefit, micropython also supports SD cards and can load it's code off there or the 4MB flash that's also on the board.

**Comment by xC0000005** [comment]
Does anyone have a lerdge-x board they've already reset the bootloader on? I think I've managed to fully decrypt the existing one (and I'm fairly certain the firmware's completely decrypted now), but obviously once I flip the bit to reset read protect and rewrite, I lose my encryption breaking ability

**Comment by mcheah** [comment]
They bothered encrypting the firmware?  That's a next step.  I'm planning on running third party firmware anyway if no one else steps up i can take the risk.
Is the bootloader encrypted too?

**Comment by mcheah** [comment]
If you can decrypt the firmware, then the easier test may be to encrypt your own firmware.  Write a blinky program, encrypt and upload it and if it works then you know you've got the right encryption scheme.  That way you don't have to touch the bootloader in case something goes wrong.  You can then read out the bootloader bit by bit and compare against the bootloader binaries.

**Comment by xC0000005** [comment]
I can. In the files section for my hackaday project, I attached a python script that decrypts whatever you hand it.  Odds are, there’s a simple way to do this that doesn’t involve a table of all 256 byte values, but since I extracted it as a byte table, I’m leaving it as such for now). 

https://hackaday.io/project/162592-lerdge-3d-printer-mainboard-hacking <https://hackaday.io/project/162592-lerdge-3d-printer-mainboard-hacking>

I may try and write a blink program later today once I extend the decrypt script to an encrypt script. The firmware seems to load at 0x8060000,. The stock bootloader is only 64k (61, but it has to end on a page boundary), so I have no idea what it’s doing with all that other space. My best guess is that all the UI bitmaps are crammed into ROM in that area.

**Comment by mcheah** [comment]
Very interesting, I will read through all the details when I have more time.  I had briefly tried altering the image files by appending zero's the the end of all the binaries.  It seemed to me that they didn't do a checksum or anything since the modified firmwares loaded just fine.  I was hoping to crash the processor by causing it to write beyond the flash ROM memory space (and thus determine the firmware memory address), but it seems they truncate or ignore the extended image.

**Comment by xC0000005** [comment]
The bootloader must be a second stage one. It loads at 0x08010000. Haven’t quite found the actual encryption code, but I will. I’m certain the decryption is working now, but I don’t know what lies before the bootloader, or after it yet.

**Comment by LoialOtter** [comment]
As soon as you're able to make a payload it'll be really easy to dump the whole image. Code running from flash has 100% access to flash for obvious reasons.

I have one board that I haven't erased and a second one I have - let me know what's needed to test.

**Comment by xC0000005** [comment]
Yes, I extracted the malyan bootloader the same way. On the board you erased, you can TRY loading this at the base of flash. If it works, you can then theoretically re-flash to the Lerdge firmware.

**Comment by xC0000005** [comment]
[decrytpedbootloader.bin.zip](https://github.com/LoialOtter/Open-Lerdge-X/files/2651707/decrytpedbootloader.bin.zip)
Let's try this again. Now you should get it.

**Comment by xC0000005** [comment]
After checking again, the decrypted bootloader definitely starts at 0x10000, so there's something earlier, I just don't know what.

**Comment by LoialOtter** [comment]
the ST parts have an interesting problem when it comes to bootloaders - the first 4k is not rewritable if readout protection is enabled. This problem generally means a dual bootloader setup. The first bootloader may be the same but compiled to operate in this range (and they may have found out about this after shipping like I did).

My solution when I found out about this was to replace a call that happened outside this 4k limit in the original bootloader with a reconfigure and restart from a 4k image. All later releases of that hardware I instead included a 4k first boot that simply reconfigured the chip for operation at 4k and restarted there.

**Comment by xC0000005** [comment]
It’s certainly possible. If I can ever get a payload to run, I can dump everything. The firmware and bootloader do NOT start the way ST’s normal code does. Neither load the SP, there is no timer configuration, so the bootloader isn’t just there to decrypt and start - it’s integral to setup. Furthermore, it’s clear there’s at least three “initial” stack pointer values in the RAM - all lerdge firmware, including the -s, use only the lower half of RAM, while the second stage bootloader sets its initial SP to near the top. 

Right now I cloned a variant and put a function to set LEDs called straight off the resetHandler, but it can’t make any calls yet, since i don’t know the state of the peripherals. I need to make  simple math based LED flasher like the turnOnFan function I used on the M200s to diagnose.

**Comment by mcheah** [comment]
I think I've got a project written for a similar STM32F4 processor to do the memory readout if that's helpful here.  I think it should be pretty easy to port over to F407 from the F446.  You'll just need to expose one of the UART interfaces.  The F07 projects are the cortex M0 version I adapted for the MPMD.

https://github.com/mcheah/stm32MemStream/tree/master/stm32MemStream

You'll need systemworkbench for stm32, and will need to update the pinmappings, but might give you a head start.  I might have some time this weekend to give it a try myself now that we know the method to encrypt/load arbitrary code.  There's a project called uBlinky that's a barebones blinky program as well.

FYI, I believe the UART/SWD pins are brought out to the 1mm header pins next to the USB port, so you should be able to do most of this with no soldering.

**Comment by xC0000005** [comment]
You are correct - the SWD pins are easily accessed via the expansion port where the network adapter plugs in. At least one UART is in the 6 pin header - an ESP8266 plugs in there for wifi printing, and that interface is purely serial.
I’m going to keep hacking until I have some control of the system.
Then, the simplest option is to write a “obliterate everything” function that resets all timers, systick, callbacks, DMA, interrrupts, etc, and then runs as if system reset had just occurred. But until I can blink an LED ont he board, I’m just working on it. I uploaded a new version to hackaday.io <http://hackaday.io/> that supports encrypting instead of just decrypting, now.

**Comment by mcheah** [comment]
Easiest would be to just disable interrupts, no?

**Comment by xC0000005** [comment]
I have learned the hard way from the M200 that doing that alone does NOT set the system up for configuration properly - if systick isn’t running wihtout any crap left over from the bootloader at pre-main, static initializers with delays will hang. Of course, I’ll try just disabling interrupts first.

**Comment by xC0000005** [comment]
Sort of basic question - which LEDs are PC7 and PC6 connected to? And is it writing high to them that turns them on or low?

**Comment by xC0000005** [comment]
Some updates (this isn't what I was looking for, but the things you find when you go looking...):
The LCD is addressed in FSMC mode. The control register is at 0x60000000, the data register at 0x60020000.
![screen shot 2018-12-08 at 8 56 38 am](https://user-images.githubusercontent.com/32139633/49688457-3b973000-fac7-11e8-9019-c5acc7b37e46.png)
I can't tell what the controller is based on the init sequence, but if nothing else, you can simply string the commands together and it will work.

I also attached a list of the pins that are configured for FMSC (that's the alternate function 0xC). 
![screen shot 2018-12-07 at 2 41 13 pm](https://user-images.githubusercontent.com/32139633/49688447-068add80-fac7-11e8-9b0b-50aac6bc69b3.png)

**Comment by xC0000005** [comment]
Also, I have zero idea now where anything is loading. I have the DLion source now thanks to someone on hackaday - DLion is the ancester of the lerdge board - none of the encryption, but you can see the structure  and nod and say "I can see how they got here from there." Particularly when you look at the font update/SPI Flash code it should be familiiar. 

The code is built with Keil, and Keil has a habit of sticking the vector table near the end. If I load the firmware at 0x10000, the vector tables correctly resolve to a set of infinite loops (which is usually the best way to track down an unknown arm vector table). The problem is, they also make some sense at 0x60000. The bootloader does NOT at 0x0000. 

I'm thinking my goal here is shifting - this bootloader is a pain in the ass to work with. What I want to do is get in and dump out the base image so people can DFU back, then we'll drop something sane on for everything else. Also, once we have the core image, it should be possible to disable flash protect and set it up. Warning - the bootloader will re-enable flash protection if it's disabled, during the update sequence.

**Comment by mcheah** [comment]
If you can isolate where the bootloader is setting the security bits, presumably you could edit the binary to NOT that no?  Or take a different route and just provide your own bootloader instead of the stock one.

What program are you using to disassemble the decrypted binary?  How are you getting the symbols for the function names?  Is that from the DLion source?

**Comment by carl1961** [comment]
Here is a Dlion manual, I downloaded before the site was took down, on page 24 starts where is shows how to burn the boot-loader, maybe it has some clues for you. I am so new to this so as of right now I have no clue what needs to be. 
[Dillon motherboard manual-ProV3.2 (1).pdf](https://github.com/LoialOtter/Open-Lerdge-X/files/2660033/Dillon.motherboard.manual-ProV3.2.1.pdf)

also Dlion Marlin GitHub of a older version may be useless

https://github.com/Pomise/Dlion-Stm32-3D

**Comment by xC0000005** [comment]
Program I'm using - BinaryNinja. Function names were added by myself, and the name is almost always evolving. Delay was 'Mess with Sys Clock." Followed by "Wait for SysClock" and then when I looked at how it was used, "Delay." The LCD code, I looked at ST's app note on interfacing with TFT LCDs and saw it used the FSMC and said "Hey, that looks familiar." If anyone has a board which they can load code on (erased), check the device ID - that will probably give us a clue to the controller for the LCD.  Yes, I could disable the function to re-enable protection. It's interesting that they thought of so many contingencies.

**Comment by xC0000005** [comment]
There's actually two different ways to address the LCD in terms of initialization - this is likely how the bootloader operations in both graphic modes. In the second, it does NOT send the wakeup sequence, but in both cases, reads the device and manufacturer ID.

I'd love to know what it reports. I may attempt to read them out via GDB (I'm not sure it is supposed to work, but we'll see). The Lerdge code expects the manufacturer bytes to add up to 0x55 (yeah, it's odd).

**Comment by xC0000005** [comment]
If you own a lerdge board you've erased, I have uploaded two files to my hacking project. One is what I see at the first 64k, the second is the decrypted bootloader update, which I swear looks like it would run at 0x10000. 

You can try and use the base bootloader. If flashed at the base of flash, you should recover the ability to reset to lerdge's released firmware. Files here:

https://hackaday.io/project/162592-lerdge-3d-printer-mainboard-hacking

**Comment by xC0000005** [comment]
@LoialOtter, have you actually succeeded running code on the lerdge you reset? If so, what clock configuration did you use? I've figured out that my code isn't executing because I never exit from the system clock config. Since I cloned my variant from a discovery board, it's highly likely it's wrong. (alternately, if anyone knows the external crystal frequency on a lerdgex, that would help, too.)

**Comment by LoialOtter** [comment]
https://github.com/LoialOtter/micropython runs fine on the board. I configured it with a 25MHz external oscillator. Check under ports/stm32/boards/LERDGE_X/stm32f4xx_hal_conf.h for hardware config.

**Comment by xC0000005** [comment]
Thank you. I’ll pay attention to how it’s configuring the clocks. The arduino code is failing to pass clock intiialization, probably because of something I did wrong. If you get a chance, could you test the decrypted bootloader (lerdgebasebootloader.bin) and see if it then allows you to reflash to lerdge’s official firmware?

**Comment by xC0000005** [comment]
Finally got control of the board by not re-initializing the HSE (or frankly, any of the clocks - ST officially warns against messing with the HSI once it's up and running, and they aren't kidding - you can kill SWD). 

https://twitter.com/AuthorJCNelson/status/1072719345078292481

Next up, getting my board variant cleaned up and posted to my github.

**Comment by carl1961** [comment]
Awesome work! @LoialOtter and @xC0000005 wish i was more knowledgeable to help. how would i properly go about preparing a x board to test? I have all the software and tools I think to do testing.

**Comment by xC0000005** [comment]
I have zero idea how to put the board into DFU mode. My next steps are going to be to build a pins.h file for the lerdge x, and maybe build a "bootloader dump" sketch. Progress on Marlin will be slow since I need USB support. That's a 1.5 feature, and the ST developer has said a PR is coming soon for it, but a lot of this is going to be "huh, we need PWM on this pin, so can't use that timer" or the like. @LoialOtter, for micropython, to get it to run with the default loader, all you should have to do is change the VECT_TABLE_ADDR (and linker script, of course) to a start offset of +0x10000, and comment out the clock config/stub it with a blank, then encrypt the resulting bin with the lerdgeencrypt2.py script.  I'm not saying you have to do this. In fact, for Marlin, I'd expect that there will be a couple of different options, one for people who just want to slap it on and try, and one for people who aren't going back, and who upgrade their bootloader to something that doesn't eat up the SPI RAM with fonts Marlin won't use. 
Edit to add "and encrypt" for the how to use.

**Comment by mcheah** [comment]
I'd guess DFU mode is that button next to the USB/expansion port.  If you hold it down while powering up it should show up in windows (you still need a USB A to A male/male cable) device manager.  If I were you I'd use the standalone UART port accessible through the expansion connector for now, and just hook up an FTDI adapter for now.  I'd also consider ditching the hardware PWM altogether and just doing it through software.  The only thing that needs microsecond tick rates is the stepper ISR, and frankly at 168MHz, you could jankily do it without a hardware timer.
There's kind of the "right way" to be developing for this board, and there's the "janky, but kind of works" way to get things running as a proof of concept.  I'm considering making a custom board through oshpark to add a microUSB, and JTAG ribbon connector to the 0.05" expansion port to make developing easier.
Does anyone know what is stored on the two external memory chips?  I think there's a 4M SPI flash and a smaller EEPROM.  Maybe that's the UI assets?

**Comment by xC0000005** [comment]
The UI Asssets are almost certainly going to SPI flash. When it fails to load the UI during bootup it logs an SPI error. There’s a picture on my hackaday project of where the two SWD pins are. If you have the USB B adapter, you could solder to those pins instead, and get the ability to remove your leads if needed.

**Comment by mcheah** [comment]
That makes sense, explains why there's a separate UI .bin file.
Yeah, they're certainly accessible, I just find when I'm screwing around with a circuit with flying wires for an extended period of time you'll eventually rip something off.  Nothing more frustrating than chasing down an intermittent bug when the SWD dips out randomly as well.
So you took the plunge and nuked the security bits now?  Or uploading encrypted binaries still?

**Comment by xC0000005** [comment]
Still uploading encrypted binaries for now. Once I get the dump sketch working, I’ll make a backup of CCRAM (it’s writing something there and actually interacting with it), SPI flash, and most of the flash space. That gives us a 100% “go back” mode for users.

Then I’m going to reset and find a bootloader that doesn’t suck.
I want to have the variant work in two ways:
Mode 1 - for users to “try” - encrypts, limited resources, etc.
Mode 2 - “Screw this, I’m moving.” - New bootloader, use SPI flash for whatever. Also, t’s nice to note that the ESP8266 webUI just slots tot he expansion port and works. You configure it as the second host and it feeds gcode.

**Comment by mcheah** [comment]
Gotcha, keep me updated.  I should have a lot more time to mess around with this for the next few weeks or so.
In my experience, most people won't bother installing anything unless it's super easy mode to install.  This of course gets way more complicated when end users need to compile to support different printer hardware.  So I'd say Mode 1 is pretty critical because people without coding experience eventually muck things up without reading instructions, blame the new software, and want to go back to stock.  I think best case scenario is if you could compile and upload from Arduino software like everybody expects for typical RAMPS boards.

**Comment by xC0000005** [comment]
Agree. The first thing out of the gate has to be something that “just works” even if it works in limited scenarios. It’s unlikely the LCD, for instance, will work at first. Oddly enough, since it costs me nothing, the secondary serial/Web printing will work straight out of the box. I know how it’s reading the touch screen as well, now. It looks like the touch screen has an pin it raises when there’s a touch. this appears to trigger a read to RAM of the X/Y, and then from there, it decides if the rectangle is within a menu border.

**Comment by mcheah** [comment]
LCD I feel will probably be a challenge unless somebody has already written some stuff for Marlin already.  For my money I think finding a way to hook up a standard reprapLCD or even a panel Due will probably be a better option than redoing the entire driver/UI of the bundled touch screen.  Interestingly though, nobody bothered using my mini delta Marlin port until I added in the LCD support (thanks to your malyanlcd code!).  So my assumption that all tinkerers are already running pronterface/repetier/octoprint was not really true.

As for the ESP8266 web server, I'm kind of starting to lean away from these kind of scenarios.  Maybe it's gotten better with Marlin 2.0, but I felt that you'd get stuttering when streaming g-code regardless of how fast the host is, especially for deltas.  Sending line-by-line in ASCII with acks does not lend to high throughput and low latency.  So the ideal scenario for me would be to stream the whole file at once, save it to the SD card, and then run off of that.  Kind of an off-topic rant, honestly just curious if anyone else has tackled this issue.

**Comment by xC0000005** [comment]
There’s someone who’s worked out how to get the MKS-ROBIN working. All of these, including DLION, are built off of ST’s “Interfaceing LCD with STM32Fx processors using FSMC” app note. So, we’ll build off their work, using the lerdge specific init sequence.

**Comment by mcheah** [comment]
In the process of mapping out the expansion connector, I intend to make a little adapter board for ST-link.  Can either of you confirm that this matches what you've seen?  One interesting thing I've noticed is that there's only one hardware UART exposed on this connector, which I presume is connected to the CH340 chip on the USB board.  But they also have some interface to the ESP8266 module in that riser connector.  @xC0000005, can you confirm the pinout of the ESP8266 module?  This seems to imply that either the USB connection or the ESP is using software serial for communication.
![lerdge expansion pinout](https://user-images.githubusercontent.com/7274015/49956014-2dd01900-fed2-11e8-9800-07386e1f18ba.png)

**Comment by mcheah** [comment]
Re: DFU Mode
I was wrong about the button in the lower left corner of the board, it's a reset button, not boot0 as I thought.  It looks like boot0 is not accessible by a connector pin.  There's a 10k resistor to ground as marked below, if you want to enable DFU loading you'll have to momentarily short it to 3.3 during powerup, probably not a great user accessible option.
![20181213_175628298_ios](https://user-images.githubusercontent.com/7274015/49958044-a84f6780-fed7-11e8-8c0c-8450c228ccc6.jpg)

**Comment by LoialOtter** [comment]
sorry I haven't been able to help at all with testing. My dev setup for this project has been disassembled for a bit and I've been swamped by bringup of a new FPGA at work (look up chronos1.4). As soon as I get a chance I'll get it set up again.

**Comment by xC0000005** [comment]
It's all good. I got the lerdge K board today from Carl. It currently crashes my mac when I plug it in, but that's because OSX and the CH340 drivers don't get along well. I'll be working on a universal bootloader dumping sketch next - I need to make this easy and near automatic, and then I'll dump the loader from the Lerdge K.  Oddly enough, the lerdge-x does NOT crash the mac. 

I compiled Marlin with a pins definition for the X, but it is failing, and I don't know why yet. There's so many moving parts right now it's entirely conceivable that it's a timer initialization issue, or USB, or something else. On my marlin fork, there's now a lerdge branch. This has the USB pull request merged into it. I'm hoping in the next week to be able to get something basic running. Not necessarily tuned or featureful, just "not crashing" would be a good start. Anyone who wants to start working on a pins_LERDGE_K file, feel free. You can always dump it off on me and I'll merge it in.

**Comment by xC0000005** [comment]
@mcheah , the ESP module is a pretty much vanilla ESP-01S.

**Comment by mcheah** [comment]
Right... But what pins does it connect to on the stm32?

**Comment by xC0000005** [comment]
I’ll see what the firmware does and get back to you.

**Comment by Dzenik** [comment]
Hi guys!
Made a gpio scheme connection of Lerdge-K. Maybe help you.
[Lerdge-K.pdf](https://github.com/LoialOtter/Open-Lerdge-X/files/2683154/Lerdge-K.pdf)

**Comment by carl1961** [comment]
@Dzenik , thanks! how was you able to do this? with a volt meter?
EDIT: just realized Open-Lerdge-X-master  is Lerdge K pin-out

**Comment by Dzenik** [comment]
[Lerdge-K.pdf](https://github.com/LoialOtter/Open-Lerdge-X/files/2684039/Lerdge-K.pdf)
Added measurement schemes for heatbed and hotend sensors.

**Comment by xC0000005** [comment]
Anyone know what the W Endstop is? X, Y, Z I get. W, I don't.

**Comment by mcheah** [comment]
I think W is typically Z-min for delta printers

**Comment by xC0000005** [comment]
Ok. I’lm leaving it out of the pins definition (note, if anyone has already build the pins.h file, you won’t hurt my feelings).

**Comment by xC0000005** [comment]
I pushed a preliminary pins and board definition for Lerdge X and K.
Please note that these are VERY preliminary, and there's good reason for it:
1. The LCD pins are addressed using FSMC. I'm not 100% sure these pins will have any meaning. The good news is that Mks-Robin needs FSMC support, too.
2. SPI - The flash ram's sitting at the other end of this, and it's not going to be used for that. Plus, Marlin's SPI config is going to need some beefing up to support this unless I'm missing something obvious.
3. SD - Again, it's on the other end of SDIO and the pins aren't negotiable, not relevant to a pins implementation. In fact, FatFS as NOT in marlin supports a SDIO implementation, but I don't think that one's in Marlin yet.  (Also, MKS-Robin also needs SDIO.)

**Comment by mcheah** [comment]
When you're talking about MKS-robin, you're talking about development on the Marlin 2.0 branch?  Or is there a separate repo for the Robin?

**Comment by xC0000005** [comment]
There’s an open PR for Mks-Robin in 2.0.

**Comment by mcheah** [comment]
the [MKS-Robin PR](https://github.com/MarlinFirmware/Marlin/pull/12650/commits) looks pretty barebones.  I wouldn't hold my breath on ST providing FSMC or SDIO hardware support to their arduino port.  Also, having worked to bring in hardware SD support to the F070 for the MPMD, I can tell you that it's going to be pretty tough to integrate it into the [Arduino SDFat library](https://github.com/greiman/SdFat).  All their drivers and examples are pretty much set up to hook into FatFS.  That's not to say that it can't be done or that you couldn't pull in FatFS into Marlin, moreso that I foresee this being a pain point.  But there's nothing to stop you from implementing SD as a SPI device as is typically done, it just significantly drags down the performance.
What about USB-MSC?  Is this something that could even be supported by Marlin?  I know that I plan on using the USB port as a virtual comm port, but in theory you could retain the USB host support and use a FTDI-like bridge on the exposed UART.

**Comment by xC0000005** [comment]
The same owner of the core wrote the STM32SDIO Arduino library, so I’m hoping. The way this would work is relatively simply - you can see that FATFS in Marlin has a chip command layer, and we abstract SDIO vs SPI there…IF I can get it working.

**Comment by LoialOtter** [comment]
The micropython code also has SDIO support but I do believe it's bitbanged. Not sure if that'll be helpful or not.

**Comment by Dzenik** [comment]
Did you watch this library?
https://github.com/stm32duino/STM32SD

**Comment by xC0000005** [comment]
That’s the one I’ve been trying to use. Today I accidentally short circuited my USB adapter for the lerdge. One resistor near the edge of the board let out its smoke. The CH340 still functions, so no telling what is broken, but I went ahead and made the ugliest diagram on earth for how the expansion header works. I’m still thinking this is a pretty much stock ESP8266-1 module. My attempts to write to the serial are not working, despite my timers being fairly accurate.

**Comment by Dzenik** [comment]
Did you write the firmware from arduino?

**Comment by xC0000005** [comment]
Yes.

**Comment by Dzenik** [comment]
[TestUART.zip](https://github.com/LoialOtter/Open-Lerdge-X/files/2687628/TestUART.zip)
Can you test this code? Encrypt and what are you doing there)
PA9 - TX 
Baudrate -115200

**Comment by xC0000005** [comment]
I'm waiting on some hardware to arrive to replace my USB->TTL converter to continue serial work. One update for the X. PC7 is the green status LED  in the corner of the board/on the LCD PCB, PC6 is the dim yellow LCD near the bed connnection.

**Comment by xC0000005** [comment]
I added the bootloader dumped from a -K board thanks to @carl1961 to my hackaday. If you have a wiped K board, you can try updating it and see if it works. I have every reason to think so, since I can see my arduino sketch sitting in ROM at the place where I figured the bootloader ended. Enjoy.

**Comment by Dzenik** [comment]
> I'm waiting on some hardware to arrive to replace my USB->TTL converter to continue serial work. One update for the X. PC7 is the green status LED in the corner of the board/on the LCD PCB, PC6 is the dim yellow LCD near the bed connnection.

Is it on the Lerdge-X board?

**Comment by xC0000005** [comment]
Yes. There’s three total LEDs. Dim yellow in the middle “bottom” near the hotend/bed connections, bright green in the upper “right” of the board (this is what flashes when the bootloader is updating firmware) and the LED on the display board, which is chained to the bright green one.

**Comment by xC0000005** [comment]
FYI, I got Serial1 sending tonight by altering the pinMap to make sure the serial pins got their alternate functions set correction. Receive doesn't work yet for some reason, and since I was testing it by sending a G28 to Marlin, I'm not sure what the reason is. Also, I have to keep the baud rate low right now - I'm working on understanding why. I'd bet money the ESP connector on the expansion board is going to turn out to be the "standard" one and when Lerdge says a regular module won't work, what they mean is "We don't want it to work." I'll figure out what's keeping RX from working eventually, but there's a number of possibilities:
1. bad code (in particular, the pinmap being wrong - this is the variant's pin map). 
2. The RX pin not being where we'd expect it to be.
3. Marlin going away but still firing the heartbeat indicator since that runs in the ISR.

We'll see.

**Comment by mcheah** [comment]
@xC0000005 ,
Trying to load arbitrary code with the stock bootloader, not getting any success.
To be clear, the application is loaded onto to flash at 0x08060000, but the vector table is loaded at 0x0801000?  I have some confusion regarding this because the stock firmware is 332k, and 0x08060000 as a start location would not fit the entire binary.  So either it gets split or there's some tomfoolery regarding the vector table.
I've tried setting up my linker script at both locations and I can't seem to get anything to run.  Can you post your blinky code/binary somewhere so I can compare against what I'm doing?

**Comment by Dzenik** [comment]
try set begining application at 0x08020000

**Comment by xC0000005** [comment]
Ok, catching everyone else up on what I told @mchea over Facebook. When you’re reverse engineering an ARM firmware, a lot of guessing about the vector table’s address is based on where jumps suddenly start producing valid code that actually does something. 0x60000 was one place where it did, but I was wrong. I’m sure it’s the 0x10000 one. You can check out the variant in my STM32duino branch and see how I have it set up.

Also, remember the LD script and the VTOR are two halves of the same coin. Your LD script should assemble the binary in a way that makes your vector table offset make sense.

**Comment by xC0000005** [comment]
I copied the Clock Init code (mostly) from MicroPython and was able to successfully initialize USB using the Arduino Core. I discovered something unpleasant about the ESP8266 board - when you plug in USB, it disables POWER to the ESP8266. I'm not sure if this happens on the expansion board or the motherboard, but Marlin supports two serial ports in 2.x and it would be nice to enable both. I haven't tried SDIO yet with the proper clock config - that'll be soon, even though Marlin doesn't properly support it.

**Comment by carl1961** [comment]
xC0000005 , that also happens on a lerdge 3d printer, if you plug the computer to your printer it disconnects the WiFi.

**Comment by xC0000005** [comment]
FYI - with proper clock initialization, the STM32SDIO library is perfectly happy to read/write the SD card. Marlin of course doesn't use this library or allow you to interface with it, but that's another layer to unravel. FSMC investigation may come sooner than I planned.

**Comment by Dzenik** [comment]
> @xC0000005 ,
> Trying to load arbitrary code with the stock bootloader, not getting any success.
> To be clear, the application is loaded onto to flash at 0x08060000, but the vector table is loaded at 0x0801000? I have some confusion regarding this because the stock firmware is 332k, and 0x08060000 as a start location would not fit the entire binary. So either it gets split or there's some tomfoolery regarding the vector table.
> I've tried setting up my linker script at both locations and I can't seem to get anything to run. Can you post your blinky code/binary somewhere so I can compare against what I'm doing?

ld script changes 
/* Highest address of the user mode stack */
_estack = 0x20010000;    /* end of RAM */

/* Specify the memory areas */
MEMORY
{
FLASH (rx)      : ORIGIN = 0x8010000, LENGTH = 312K

**Comment by Dzenik** [comment]
Happy New Year!

**Comment by xC0000005** [comment]
Happy New Year to you as well!

Aand in case I didn’t answer your question from earlier in the thread, you want your LD script and your VectablOffset to both use 0x10000.
You can pick up working ld scripts and vector table offset definitions from my ST arduino core github branches.

**Comment by Dzenik** [comment]
Yes, I changed the ld script and VECT_TAB and the program did not start. Not yet changed _estack and MEMORY order array

**Comment by xC0000005** [comment]
Are you running with Lerdge’s bootloader?

**Comment by Dzenik** [comment]
Yes

**Comment by xC0000005** [comment]
Does the lerdgecrypt.py script successfully encrypt the binary?

**Comment by Dzenik** [comment]
Yes

**Comment by xC0000005** [comment]
Ok. Build a blink binary (blink both LEDs. :)) and send me both the unencrypted one for me to look at in binary ninja and the encrypted one for me to load on my board. We’ll make sure your toolchain works and then you can get fancy.

**Comment by Dzenik** [comment]
[Debug.zip](https://github.com/LoialOtter/Open-Lerdge-X/files/2718670/Debug.zip)

**Comment by xC0000005** [comment]
Ok, Binary Ninja says it’s a fairly normal blink. Testing on the lerdge K board will have to wait until this evening, as it’s time for me to head to work. I do almost everything right now on the Lerdge-X. Other than dumping the bootloader from the K, I haven’t tinkered with it while I work out the gritty details of Marlin, SDIO, and the LCD, so it’s entirely possible there’s something different about the K.

Which pins are you pulsing for the LED?

**Comment by Dzenik** [comment]
PB6
Led marked on board Lerdge-K

**Comment by xC0000005** [comment]
That looks right. I’ll test when I’m home and see if I can figure out what’s going on.

**Comment by mcheah** [comment]
FYI, I had to change the  .ld script and use @xc0000005's startup.s code to get things to boot.  I also discovered that changing the MODER register would lock things up for some reason, haven't really investigated further.

@dzenik, are you using the stm32 arduino toolchain or the stm32 hal directly?

**Comment by xC0000005** [comment]
What startup code? I dropped the custom.s a while back. It’s literally using the HAL startup. I would guess whatever you were using doesn’t load the SP, which the bootloader does not.

**Comment by Dzenik** [comment]
> FYI, I had to change the .ld script and use @xC0000005's startup.s code to get things to boot. I also discovered that changing the MODER register would lock things up for some reason, haven't really investigated further.
> 
> @Dzenik, are you using the stm32 arduino toolchain or the stm32 hal directly?

HAL from STM32 Workbench

**Comment by Dzenik** [comment]
We can use @jmz52 code for LCD. Need proper config FSMC and use initialize code for our display.

**Comment by jmz52** [comment]
@xC0000005, I've added support for SDIO for Robin board. Just need a day or two for minor cleanup before sending PR.
Hardware initialization is handled by stm32duino, but rest of the code should be more or less reusable for F4 with different framework.

**Comment by xC0000005** [comment]
How did you abstract away Marlin’s direct use of the SD card?  I’m currently using ST’s arduino library for SDIO, which works, but I was just beginning to figure out how to substitute it in for Marlin’s pretty much direct SD access.

**Comment by jmz52** [comment]
I've uploaded preliminary code to https://github.com/jmz52/Marlin/tree/STM32F1-SDIO-support-rc0

Marlin/src/sd/SdVolume.h: 
With "SDIO_SUPPORT" defined the #include "Sd2Card.h" is replaced by #include "Sd2Card_sdio.h"
Sd2Card class is redefined with the only 4 methods used by Marlin - init, writeBlock, readBlock and cardSize(returns number of 512-bytes blocks)

These methods relies on HAL-specific functions to do actual work:
bool SDIO_Init(void);
bool SDIO_ReadBlock(uint32_t block, uint8_t *dst);
bool SDIO_WriteBlock(uint32_t block, const uint8_t *src);
uint32_t SDIO_GetCardSize(void);

You need two defines to make it work (will be changed to one define)
#define SDSUPPORT
#define SDIO_SUPPORT

Please note that stm32duino uses different SDIO registers structure than CubeMX (STM32 HAL libraries).
stm32duino defines RESP as an array when CubeMX defines it as four uint32_t variables.

**Comment by jmz52** [comment]
And it is looks like the cardSize() method is not being used after all, so only three methods are needed.

**Comment by xC0000005** [comment]
I’d actually looked at the block interface and thought about modifying it there. I’ll give it a spin on the unified HAL (which is where, eventually, most stuff will land). Right now, STM32DUINO is slightly ahead in terms of features, and will stay that way until the 1.5 release gives USB to most of the boards.

**Comment by jmz52** [comment]
PR for SDIO support submitted.
https://github.com/MarlinFirmware/Marlin/pull/12782

I did not find a good place to add  a dependency, so both defines are still needed to make it work.
`#define SDSUPPORT`
`#define SDIO_SUPPORT`

**Comment by mcheah** [comment]
@xC0000005,
Some updates, I managed to dump out the bootloader over UART with my own app.  The binaries don't match your lerdge x bootloader binary, but I suspect this is because I updated my bootloader to [V1.0.4](http://www.lerdge.com/Private/Files/20181016095217426%E2%88%AELerdge_X_Bootloader.zip) as provided by Lerdge before starting this endeavor.  Interestingly it matches the [Lerdge K bootloader you posted](https://cdn.hackaday.io/files/1625926956336128/lergdekloader.bin) fairly well, but not exactly.  It says in plaintext in the binary the version number.

By comparing against the bootloader binary provided by lerdge, it provides some interesting insights.  It seems that 0x07D4 - 0xB020 maps out to 0x0000 to 0xA84C.  Beyond 0xA84C I get bytes filled with 0xFF's.  It's unclear where/what the rest of the bootloader binary provided by lerdge does.  My suspicion is that it gets loaded onto SDRAM, or perhaps some of the assets get loaded onto the SPI flash.  But for the purposes of restoring the device to stock via SWDIO, it probably doesn't matter, we have our own backups now.  Another nice thing I confirmed is that the entire application binary gets loaded directly to 0x80010000 as is, which is really the more important thing for running custom code.

The next steps for me are to nuke the security bits and start poking around at the register settings.  I need to adapt the drivers for my custom stm32 Marlin fork.  Hopefully this will help flesh out some of the platform specific issues so that they can be integrated into the Marlin 2.0 port.

[lerdgexboot_1.0.4.zip](https://github.com/LoialOtter/Open-Lerdge-X/files/2723040/lerdgexboot_1.0.4.zip)

**Comment by xC0000005** [comment]
Fantastic. Judging from my reverse engineering of the bootloader, assets get loaded primarily to SDRAM. There’s FATFS support for SPI, and I think they use that. My time on tis project is going to be limited due to day-job commitments for a bit. The SDIO hal abstraction PR is the right way to incorporate it, I just can’t get the raw SDIO block commands to work at the moment. If you flash bootloader and firmware *together* I don’t believe it re-protects flash.  I’ll upload my lerdge reset script to Hackaday for you shortly.

**Comment by xC0000005** [comment]
I uploaded my opencd cfg I use to reset the lerdge to Hackaday. Also @dzenik has managed to init and fill the LCD(!).

**Comment by mcheah** [comment]
I'm using the stm32-utility for windows, I think it's just a GUI-friendly way of doing the openocd stuff.  I believe the bootloader (or at least on my system) re-enables the option bytes regardless of whether it is flashing a new firmware.  I tried patching the bootloader binary by replacing the 0x08192A3B and 0x4C5D6E7F flash key constants with 0x00000000, which correctly prevents the option bytes from being overwritten, but then the bootloader hangs in an infinite loop, either because of a fault when attempting to write to the flash option bytes or a status check to make sure the bits were set.  

Not quite as easy as I was hoping.  I may try my hand at some disassembly.  In theory once you nuke the security bits and reflash the micro you should be able to single-step to see where those memory accesses take place and remove those instructions.  The easier strategy is probably to just dump the peripheral registers and start developing without the bootloader for now.

**Comment by xC0000005** [comment]
I’ve done exactly this, and the answer is, I can’t find a place where it sets read protection level 2 (0x55).
Here’s the check for whether or not it’s unprotected:


I’m not saying it doesn’t exist, just that I can’t find it yet. Hang on, going to run this through retdec.

**Comment by xC0000005** [comment]
I can’t get ret-dec to run on my mac right now. If you want to download it and run it, tell me where it’s reseting the option bytes and I’ll use BinaryNinja to blast it (or you can).
https://github.com/avast-tl/retdec <https://github.com/avast-tl/retdec>

**Comment by Dzenik** [comment]
> I've uploaded preliminary code to https://github.com/jmz52/Marlin/tree/STM32F1-SDIO-support-rc0
> 
> Marlin/src/sd/SdVolume.h:
> With "SDIO_SUPPORT" defined the #include "Sd2Card.h" is replaced by #include "Sd2Card_sdio.h"
> Sd2Card class is redefined with the only 4 methods used by Marlin - init, writeBlock, readBlock and cardSize(returns number of 512-bytes blocks)
> 
> These methods relies on HAL-specific functions to do actual work:
> bool SDIO_Init(void);
> bool SDIO_ReadBlock(uint32_t block, uint8_t *dst);
> bool SDIO_WriteBlock(uint32_t block, const uint8_t *src);
> uint32_t SDIO_GetCardSize(void);
> 
> You need two defines to make it work (will be changed to one define)
> #define SDSUPPORT
> #define SDIO_SUPPORT
> 
> Please note that stm32duino uses different SDIO registers structure than CubeMX (STM32 HAL libraries).
> stm32duino defines RESP as an array when CubeMX defines it as four uint32_t variables.

What tool are you using? I revised STM32duino and stevstrong/Arduino_STM32 but did not find the right one to compile. Specify where I can download the correct one so that your code is compiled.

**Comment by xC0000005** [comment]
I’ve managed to get a file list using @jmz52’s SDIO abstraction method. I’ve got a ton of niggling things to fix and some code to consolidate, but it should be possible to do a test print soon.

**Comment by jmz52** [comment]
I am using PlatformIO with whatever packages it downloads by itself.
You can find environment parameters in [env:mks_robin] section of platformio.ini file.

If I am reading the manifest file correctly current version is http://dl.platformio.org/packages/framework-arduinoststm32-3.10400.181126.tar.gz

**Comment by xC0000005** [comment]
So STM32DUINO’s F4 core. I’m building in the Arduino IDE against ST’s core right now, but the official core lacks some key features (USB) that will be required for it to catch up.

**Comment by mcheah** [comment]
I think I got it, I used cutter (radare2) to disassemble.  There's a branch where it checks for 0xAA in the RDP option bytes, if you bypass it, it'll leave it alone.
![rdp branch](https://user-images.githubusercontent.com/7274015/50657878-aec88080-0f65-11e9-9623-295d180b8240.png)
![rdp change](https://user-images.githubusercontent.com/7274015/50657882-b1c37100-0f65-11e9-925c-225a29831763.png)




[lerdgexboot1.0.2.patched.zip](https://github.com/LoialOtter/Open-Lerdge-X/files/2725094/lerdgexboot1.0.2.patched.zip)

**Comment by xC0000005** [comment]
Try this.

**Comment by mcheah** [comment]
It appears the application binary also resets the security bits.  It's almost like they don't want us doing this.  The same patch applies to the application binary.  Also, for whatever reason just flashing the app at 0x08010000 would not start the program right.  I had to re-load an encrypted binary through the SD bootloader in order to get the app to start up again.  When I dumped the flash, it looks the same, so I suspect there are some UI assets on the SPI flash that got screwed up when I was mucking around with the bootloader.

Now I can debug and poke around the app firmware to better understand how they're utilizing the board.  It looks like the bootloader does not use the HSE crystal, but the app firmware does.  It also looks like the SDIO peripheral is not initialized in anyway, which would imply that they're bit-banging the SD card.  It's possible that part of the reason the SDIO drivers aren't working is because they're not hooked up right or not all the necessary connections are hooked up.  The only SPI interface initialized is SPI1, which is hooked up to the touch-screen.

[lerdgexapp1.0.7.patched.zip](https://github.com/LoialOtter/Open-Lerdge-X/files/2725342/lerdgexapp1.0.7.patched.zip)

**Comment by xC0000005** [comment]
I have SDIO working now. The key was using the micropython timer init code.

**Comment by jmz52** [comment]
I've added three on-screen buttons to mimic functionality of encoder Robin does not have.
Code is POC-quality and I do not plan to submit PR this feature in it's current state, but some bits might be used for proper UI.
https://github.com/jmz52/Marlin/tree/MKS-Robin-Touch-POC
Screenshot: https://ibb.co/p1Z9Q2K
Please pay no attention to green color - it is just a "print completed" indication from PRINTER_EVENT_LEDS feature.

**Comment by jmz52** [comment]
> What about USB-MSC? Is this something that could even be supported by Marlin?

Is not this is what "SHARED_SD_CARD" feature of HAL_LPC1768 is for?

**Comment by Dzenik** [comment]
> I am using PlatformIO with whatever packages it downloads by itself.
> You can find environment parameters in [env:mks_robin] section of platformio.ini file.
> 
> If I am reading the manifest file correctly current version is http://dl.platformio.org/packages/framework-arduinoststm32-3.10400.181126.tar.gz

My PlatformIO not see HAL definition
`d:\1\marlin\marlin\src\hal\hal_stm32f4\watchdog_STM32F4.h:24:8: error: 'IWDG_HandleTypeDef' does not name a type`
although in the includes path they are present

**Comment by mcheah** [comment]
I got a bare-metal version of my stm32 Marlin port working on Lerdge X.  Motion, heating, and mesh bed leveling is working as expected.  Still have to add SD, LCD, USB, and EEPROM support, although with an FTDI board connected to UART it's probably not worth the effort to implement USB CDC.  Not quite ready for prime time, but it's a good proof of concept.

https://github.com/mcheah/Marlin4MPMD/tree/lerdgex

**Comment by xC0000005** [comment]
PIO recently cut over support to the new ST Arduino core. You can use an older version of the platform to conftinue with STM32duino.

**Comment by xC0000005** [comment]
The STM32SD library or BSP middleware works fine. I created a binary that dumps the bootloader now so I don’t have to dump it via UART.

**Comment by xC0000005** [comment]
Also - for touch screen, https://platformio.org/lib/show/542/XPT2046_Touchscreen/ <https://platformio.org/lib/show/542/XPT2046_Touchscreen/> or the BSP code seems to work.

**Comment by jmz52** [comment]
What LCD controllers (or their IDs) are used on Lerdge boards?
I've just found that one of older version of MKS Robin TFT uses ILI9328 controller with 16-bit data for writing registers. And with completely different commands codes and parameters.
ILI9341 and ST7789 uses 8-bit data for writing registers, 16-bit data are only used when writing data to GRAM.
And 16-bit data for writing registers does not fit well into writeByte() function of u8g communication layer.
I am going to rework communication layer for TFT screens and I would like to check datasheets of controllers it will need to support someday.

**Comment by Dzenik** [comment]
My display is 0x04 0x04 0x04. Someone says that the HX8357 is installed, and someone that ST7789. But I am sure in one thing that data registers are 8-bit.

**Comment by Dzenik** [comment]
static const uint8_t LERDGE_init_sequence[] = { // 0x0404 
  U8G_ESC_ADR(0),
  0x10,
  U8G_ESC_DLY(10),
  0x01,
  U8G_ESC_DLY(100), U8G_ESC_DLY(100),
  0x11,
  U8G_ESC_DLY(120),
  0xF0, U8G_ESC_ADR(1), 0xC3,
  U8G_ESC_ADR(0), 0xF0, U8G_ESC_ADR(1), 0x96,
  U8G_ESC_ADR(0), 0x36, U8G_ESC_ADR(1), 0x28,
  U8G_ESC_ADR(0), 0x3A, U8G_ESC_ADR(1), 0x55,
  U8G_ESC_ADR(0), 0xB4, U8G_ESC_ADR(1), 0x01,
  U8G_ESC_ADR(0), 0xB7, U8G_ESC_ADR(1), 0xC6,
  U8G_ESC_ADR(0), 0xC1, U8G_ESC_ADR(1), 0x15,
  U8G_ESC_ADR(0), 0xC2, U8G_ESC_ADR(1), 0xAF,
  U8G_ESC_ADR(0), 0xC3, U8G_ESC_ADR(1), 0x09,
  U8G_ESC_ADR(0), 0xC5, U8G_ESC_ADR(1), 0x22,
  U8G_ESC_ADR(0), 0xC6, U8G_ESC_ADR(1), 0,
  U8G_ESC_ADR(0), 0xE8, U8G_ESC_ADR(1), 0x40,
  U8G_ESC_ADR(0), 0x8A, U8G_ESC_ADR(1), 0, 0, 0, 0x29, 0x19, 0xA5, 0x33,
  U8G_ESC_ADR(0), 0xE0, U8G_ESC_ADR(1), 0xF0, 0x04, 0x08, 0x09, 0x08, 0x15, 0x2F, 0x42, 0x46, 0x28, 0x15, 0x16, 0x29, 0x2D,
  U8G_ESC_ADR(0), 0xE1, U8G_ESC_ADR(1), 0xF0, 0x04, 0x09, 0x09, 0x08, 0x15, 0x2E, 0x46, 0x46, 0x28, 0x15, 0x15, 0x29, 0x2D,
  U8G_ESC_ADR(0), LCD_COLUMN, U8G_ESC_ADR(1), 0x00, 0x00, 0x01, 0x3F,
  U8G_ESC_ADR(0), LCD_ROW,    U8G_ESC_ADR(1), 0x00, 0x00, 0x01, 0xDF,
  U8G_ESC_ADR(0), 0x20,
  U8G_ESC_ADR(0), 0x53, U8G_ESC_ADR(1), 0x24,
  U8G_ESC_ADR(0), 0xF0, U8G_ESC_ADR(1), 0x3C,
  U8G_ESC_ADR(0), 0xF0, U8G_ESC_ADR(1), 0x69,
  U8G_ESC_DLY(100),
  U8G_ESC_ADR(0), 0x29, LCD_WRITE_RAM,
  U8G_ESC_END
};

**Comment by xC0000005** [comment]
I will try to get MCU friend running and see what it says about manufacturers and ID.

**Comment by jmz52** [comment]
@Dzenik, id 0x04 0x04 0x04 is what I am getting when display is not connected - 0x04 is "Read display identification information" register, so FSMC simply returns me last written value.
Could you please check if id will change when display is not connected?

**Comment by carl1961** [comment]
will the DLION LCD code help yall?
https://drive.google.com/file/d/1gIYMQCTqnYmQm7kG0WSCS2Zjvvn6pq-C/view?usp=sharing

**Comment by xC0000005** [comment]
His code seems to work when writing.

**Comment by Dzenik** [comment]
@jmz52 you're right. with the display off the same code

**Comment by Dzenik** [comment]
@jmz52 in the bootloader there is a check for the display type, if the code is 0x55 one initialization otherwise the other

**Comment by jmz52** [comment]
Just FYI - STM32 core was added to PlatformIO's arduino framework.
If board's "variant" is listed in .platformio\packages\framework-arduinoststm32\STM32\variants then STM32 core is used.
There are no STM32F407Zx variants, so I've used closest available board to test if code will be compiled at all.
There were some errors in SDIO and u8glib, but after disabling TFT and SDIO I was able to compile
compile LerdgeXAndK branch using following environment settings:
> [env:lerdge]
> platform      = ststm32
> framework     = arduino
> board         = **`disco_f407vg`**
> build_flags   = ${common.build_flags} -DHAVE_HWSERIAL1
> lib_deps      = ${common.lib_deps}
> lib_ignore    = Adafruit NeoPixel, c1921b4, TMCStepper, U8glib-HAL, TMC26XStepper
> src_filter    = ${common.default_src_filter} +<src/HAL/HAL_STM32>
> monitor_speed = 250000

**Comment by xC0000005** [comment]
I’ve built variants and am currently building from Arduino’s IDE, but I’ll eventually switch back to PIO. Also, I want to use the same method you did to run the produced firmware through the lerdge encryption script I have (right now, I do it manually).
I have SD and LCD working in a Marlin 2.0 branch. The only problems thus far are:
1. The code’s hacked up. I’ll start cleaning it soon and preparing for a PR.
2. I am not actually sure how the scaling works. I made a U8G 480x320 file and changed all the names, but either I haven’t had enough coffee or I’m just not grasping how the 24x320 actually does the scaling up. I need to essentially triple each pixel where the Robin doubles it.

You can add the lerdge variants from the poorly-named ST-3DP_Eval branch in my Arduino github, or if you want TFT/SDIO, just add the -Ds to simulate the hal_conf file. 

Something still isn’t working with my UG8 library - I am manually calling the init function right now, and once I do, it all works.

**Comment by jmz52** [comment]
File u8g_dev_tft_480x320_upscale_from_128x64.cpp
`uint16_t buffer[256];` should be `uint16_t buffer[384];` for 3x scaling
`buffer[k++] = c; buffer[k++] = c;` this double pixels horizontally, add one more `buffer[k++] = c;` to get them tripled
`for (k = 0; k < 2; k++) {` double lines vertically, `k < 3` to triple it
and you need to add
`u8g_WriteSequence(u8g, dev, 128, (uint8_t*)&(buffer[256]));`
`u8g_WriteSequence(u8g, dev, 128, (uint8_t*)&(buffer[320]));`
at the end of this cycle to write extended buffer
128 here is the number of bytes, so each call writes 64 pixels.

**Comment by mcheah** [comment]
My stm32 branch includes parameterized scaling based on jmz's code.
https://github.com/mcheah/Marlin4MPMD/blob/lerdgex/MPMD_3dPrinter/Marlin/ultralcd_ili3941_u8glib_stm32.h

**Comment by xC0000005** [comment]
Ok, now I understand it. Once I get everything functioning without the init hacks, I’ll revisit this. I have SD and LCD working, it’s time to put this in a printer and test it. @jmz’s SDIO SD Card support has also shown me how I intend to allow Marlin to store graphics in the SPI flash. I’ll just build a block interface for the SPI flash and let the code treat the bitmaps/logos/screens like files.

**Comment by mcheah** [comment]
I was curious about that.  Are you going put FAT on the SPI flash?  Or just hard-code block locations for the UI assets?

**Comment by xC0000005** [comment]
First I’m going to dump the FAT table that’s on the SPI flash and look at what it’s doing already. Then probably put FAT on the SPI flash.

**Comment by xC0000005** [comment]
FYI, the SPI stuff may be using SPIFFS - it would match how they’re both using file names and offsets. In essence, they have “volumes” packed into SPI flash. The volume offsets are hardcoded. The names within the volumes determine where everything is.  
https://docs.espressif.com/projects/esp-idf/en/latest/api-reference/storage/spiffs.html <https://docs.espressif.com/projects/esp-idf/en/latest/api-reference/storage/spiffs.html>

**Comment by mcheah** [comment]
Interesting.  I was gonna say, full FAT seems overkill

**Comment by xC0000005** [comment]
SPIFFS doesn’t support directories, but other than that, it’s a block layout. I’m not convinced the savings make sense. Have you tried getting either the I2C flash or the NOR SPI flash working? I want to use the I2C flash as “EEPROM” and the SPI flash for UI assets.

**Comment by mcheah** [comment]
I have not,
my firmware saves settings to the SD card anyway.  I'm curious why they need an I2C flash if they can just save to the internal flash.  Maybe endurance concerns?  Maybe it has something to do with the power loss recovery.  Does Marlin support power loss recovery?  Or was that an add-on prusa made for their fork?

**Comment by xC0000005** [comment]
The I2C package is specifically marketed as a wear resistant, lower energy eeprom replacement, suitable (according to marketing) for live writes. The NOR ram is slower, made for DMA transfers to the LCD. In a lot of ways, it’s the textbook (well, app note) ST design for interop with a TFT LCD.

**Comment by mcheah** [comment]
makes sense.  I can try pulling data off the I2C flash if you want to focus on the NOR one

**Comment by xC0000005** [comment]
I’m not sure there’s anything “good” there, but eventually I’d like to enable it for EEPROM usage, and also just in case Lerdge does something sneaking in whatever their next revision is and puts keys in there.

**Comment by mcheah** [comment]
I was able to initialize the I2C driver and read the contents of the serial FRAM.  I've attached the dump.  It uses standard EEPROM command format <I2CAddr+RWbit> <A15:8> <A7:0> <D7:0>.  I used the HAL_I2C_Mem_Read/HAL_I2C_Mem_Write functions, library like [extEEPROM](https://github.com/JChristensen/extEEPROM) should work fine.  See [Here](https://github.com/mcheah/Marlin4MPMD/blob/a91d9335cccc31ad34d41dad656350a9c26c7434/MPMD_3dPrinter/Marlin/Marlin_main.cpp#L885) for the test code

I2C address is 0xA0 (including R/W LSB)

[EEPROM_dump.txt](https://github.com/LoialOtter/Open-Lerdge-X/files/2747374/EEPROM_dump.txt)

**Comment by xC0000005** [comment]
Looking at the code disassembly and the output, I’d guess the only thing they’re storing there is some kind of print recovery and a few settings. Enabling EEPROM should be a simple as defining the I2C_EEPROM support. Wow.

**Comment by xC0000005** [comment]
Simply adding The I2C EEPROM flag to the lerdge pins and enabling EEPROM in the configuration has made it work for both X and K.

**Comment by jmz52** [comment]
@xC0000005 I've reworked TFT and touch code in https://github.com/jmz52/Marlin/tree/MKS-Robin-Touch-POC
During initialization "com" function exposes IO function to "dev" function via 
`struct LCD_IO {`
`  uint32_t id;`
**`  void (*writeRegister)(uint16_t reg);`**
**`  uint16_t (*readData)(void);`**
**`  void (*writeData)(uint16_t data);`**
**`  void (*writeMultiple)(uint16_t data, uint32_t count);`**
**`  void (*writeSequence)(uint16_t *data, uint16_t length);`**
`  void (*setWindow)(uint16_t Xmin, uint16_t Ymin, uint16_t Xmax, uint16_t Ymax);`
`};`

These IO functions are used directly by "dev" to overcome limitations of u8g and to get significant performance boost.
Initialization sequences were reworked to support 16-bit data required for ili9328 controller.
Because of these 16-bit data requirements there is no backward compatibility with original u8g 'com' functions calls.

From you point of view what is the preferable design for 'com' function - keep IO structure to overcome u8g limitation or recreate 'com' function from scratch and use completely new data, methods, parameters and data formats?
Example:
u8g 'com' flow: set 'register select' line; write byte; reset 'register select' line; write byte
new 'com' flow: write register; write data

**Update:** Also touch screen calibration was reworked to support MKS Robin TFT v1.1
Touch calibration tool for Robin's 320x240 screen - https://github.com/jmz52/MKS-Robin-toolkit/tree/master/lcd_and_touch
Some bits of it might be useful for you too.

**Comment by xC0000005** [comment]
My opinion, worth exactly what you paid for it, is that the below works perfectly well. The question I have is, UG8 seems to do some pin manipulation as well, and I’m curious if we can abstract away SPI versus FSMC. I have an MKS ROBIN now (thanks to an aliexpress christmas sale, and I’ll try getting it running using the HALSTM32. 

My code currently does not use DMA for SDIO, and the LCD code works, but doesn’t initialize right, which is why there’s no PR for it. Eventually I’m hoping that ST’s arduino core will catch up, because it supports pretty much every version of ST’s processors, but for now, I think you’re doing the right thing by getting it working using STM32DUINO’s support.

You can see my very much work in progress code here:
https://github.com/xC0000005/Marlin/tree/LerdgeXAndK <https://github.com/xC0000005/Marlin/tree/LerdgeXAndK>
It’s derived from the BSP layer, but fixes a few bugs with read block. I’m hooking the Lerdge-X to a printer this weekend. Once I get a few test prints running, I’ll do a PR for pins first, followed by getting an SDIO PR for the HALSTM32, and then LCD/Touch.

**Comment by jmz52** [comment]
SPI vs. FSMC can be done with two extra calls like 'begin' and 'end' that will be empty for FSMC and set/reset CS pin for SPI. A0 a.k.a Register Select pin manipulation fits perfectly to writeRegister and writeData calls. And the 'reset' call is universal.

My only concern here is 16-bits vs 8-bits data over SPI. There are few ways to deal with this problem, 
but SPI  version of TFT screen is needed to test exact controller's behavior.

**Comment by mcheah** [comment]
For ESP8266 there's goofy stuff like 9-bit SPI for 18-bit color mode too.  Which begs the question, how are the different color formats being handled?  Is this something that can be abstracted or does that live at another level?

**Comment by xC0000005** [comment]
I would imagine there’d be a pipeline prep for specific boards, in particular if we use an SPIFFS layout and have to pack the assets anyway.

**Comment by jmz52** [comment]
@xC0000005, I think I've found the math behind decryptionTable.
Two function below provides the same results as 'encryptionTable.get()' and 'decryptionTable.get()'
`def encryptByte(byte):`
`  byte = 0xFF & ((byte << 6) | (byte >> 2))`
`  i = 0x58 + byte`
`  j = 0x05 + byte + (i >> 8)`
`  byte = (0xF8 & i) | (0x07 & j)`
`  return byte`

`def decryptByte(byte):`
`  i = 0xA8 + byte`
`  j = 0x02 + byte + (i >> 8)`
`  byte = (0xF8 & i) | (0x07 & j)`
`  byte = 0xFF & ((byte >> 6) | (byte << 2))`
`  return byte`

**Comment by xC0000005** [comment]
This is most likely right. If you don’t mind me asking, how did you figure it out? I had it on my “todo” list for when I have everything working, but once I got a complete 256 byte table dumped out, it was never as important. Looking down the road to STSTM32 support for this board and building in Platformio, these functions would make it simple to do the same thing you’ve done with the mks-robin boards - have the firmware encrypted after build.

Again, thank you so much for looking at this. The documentation is every bit as important as getting any one build working. If we document the hardware and encryption, there’s no reason Marlin, Repetier, Klipper, whatever can’t run on this.

**Comment by jmz52** [comment]
I've looked for patterns in binary numbers format of decryptionTable, to find what bits of value are flipped by each bit of key (first guess was that there is a set of XOR operations on 0xAA byte). This gave me a set of 8 rules that successfully decoded decent chunk of decryptionTable. Then I've tried to create finer rules to cover remaining values and found that XOR is not the right tool to use (rules became more complex than patterns visible in HEX view).
So I've created 8 maps - one for each bit of decryptionTable's values to look at these pattern in HEX view. These maps turn out to be the key (example: bit 0x04 is set for odd keys when key is less than 0x58 and for even keys when key is from 0x58 to 0xFF).
While working with these maps I've found that starting point is 0x58 and that lower three bits in key have offsets shifted at 0x58.
First working version had 8 rules - one for each bit.
`for i in range(256):`
`  if i < 0x58:`
`    j = i + 0xA8`
`    k = j + 2`
`  else:`
`    j = i - 0x58`
`    k = j + 3`
`  a = 0x00`
`  a |= 0x01 if j & 0x40 else 0`
`  a |= 0x02 if j & 0x80 else 0`
`  a |= 0x04 if k & 0x01 else 0`
`  a |= 0x08 if k & 0x02 else 0`
`  a |= 0x10 if k & 0x04 else 0`
`  a |= 0x20 if j & 0x08 else 0`
`  a |= 0x40 if j & 0x10 else 0`
`  a |= 0x80 if j & 0x20 else 0`

Then I've reverted these rules to create encryption function with another 8 rules and reworked code to make it as short as possible. Never expected to end with just 4 lines though.

**Comment by xC0000005** [comment]
I don't know if @mcheah already found this but I'd bet so: The PDF has the step and dir pins for the extruder on the lerdge x reversed, as I found out today while attempting a test print. (On the other hand, hooray for finally getting to the point of a test print).

**Comment by mcheah** [comment]
You know, I noticed that at one point, but I assumed I had fat-fingered it into the header file.   Good to know I'm not crazy

**Comment by xC0000005** [comment]
@LoialOtter, I have a working extruder with these pins:
#define E0_STEP_PIN        PA4
#define E0_DIR_PIN         PA3
#define E0_ENABLE_PIN      PA5

**Comment by Dzenik** [comment]
Help guys figure it out. I tried all the ones. Simple blink code does not work.
PlatformIO - tried the new and old platform.
Arduino - I tried stm32duino, although they added support for the F407Z, but it still doesn't work.

**Comment by Dzenik** [comment]
I took the settings from the board @xC0000005  and it worked!

**Comment by jmz52** [comment]
PlatformIO has updated platforms\ststm32 and packages\framework-arduinoststm32.
The arduinoststm32 framework now includes STM32 code v1.5

With a little 'extra_script' magic I was able to compile Marlin using custom 'board' and 'variant' definitions
>[env:black_stm32f407ve]
>platform = ststm32
>framework = arduino
>board = blackSTM32F407VET6
>extra_scripts = pre:buildroot/share/PlatformIO/scripts/black_stm32f407vet6.py
>...

PlatformIO uses 'board' definition from `buildroot/share/PlatformIO/boards/blackSTM32F407VET6.json`
>[platformio]
>boards_dir  = buildroot/share/PlatformIO/boards

And 'variant'-specific code from `buildroot/share/PlatformIO/variants/MARLIN_F407VE`
The 'variant'-specific code is copied to packages/framework-arduinoststm32/variants by black_stm32f407vet6.py 'extra-script'

The black_stm32f407vet6.py script
>import os,shutil
>from SCons.Script import DefaultEnvironment
>from platformio import util
>
>env = DefaultEnvironment()
>platform = env.PioPlatform()
>board = env.BoardConfig()
>
>FRAMEWORK_DIR = platform.get_package_dir("framework-arduinoststm32")
>CMSIS_DIR = os.path.join(FRAMEWORK_DIR, "CMSIS", "CMSIS")
>assert os.path.isdir(FRAMEWORK_DIR)
>assert os.path.isdir(CMSIS_DIR)
>assert os.path.isdir("buildroot/share/PlatformIO/variants")
>
>mcu_type = board.get("build.mcu")[:-2]
>variant = board.get("build.variant")
>series = mcu_type[:7].upper() + "xx"
>variant_dir = os.path.join(FRAMEWORK_DIR, "variants", variant)
>
>source_dir = os.path.join("buildroot/share/PlatformIO/variants", variant)
>assert os.path.isdir(source_dir)
>
>if not os.path.isdir(variant_dir):
>&nbsp;&nbsp;&nbsp;&nbsp;os.mkdir(variant_dir)
>&nbsp;&nbsp;&nbsp;&nbsp;source_files = os.listdir(source_dir)
>&nbsp;&nbsp;&nbsp;&nbsp;for file_name in source_files:
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;full_file_name = os.path.join(source_dir, file_name)
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if os.path.isfile(full_file_name):
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;shutil.copy(full_file_name, variant_dir)


Tested on my nice little (100x81mm) [shield for black stm32f407ve](https://ibb.co/vsjkkDx)

**Comment by jmz52** [comment]
I think I've nailed down the SDIO stability problem.
Sometimes there is a rogue 0xFFFFFFFF reading before actual data.
This results in 4 bytes offset in block's data and lost 4 bytes in SDIO FIFO buffer.
These lost bytes corrupts next reading and so on an so forth.

A patched version for HAL_STM32F1 - https://github.com/jmz52/Marlin/commit/e1a29e3970a563866111aab9358c8f3a40c2f8a2#diff-c12a0ff587f1645efeb7f05563654fbf

**Comment by jmz52** [comment]
Marlin now supports SDIO on STM32F4 boards (HAL_STM32)
https://github.com/MarlinFirmware/Marlin/pull/16756

**Comment by jmz52** [comment]
TFT screen and touch panel support for HAL STM32.
Currently for 320x240 screens but can be extended to 480x320 with appropriate MarlinUI implementation.
Also some extra coding is needed for appropriate FSMC and SPI initialization on STM32F4
https://github.com/MarlinFirmware/Marlin/pull/18130

**Comment by xC0000005** [comment]
The Chitu board is also a 103Z (I see you added support for it). I’ll have to see how to get the chitu environment to build using the STM32 HAL, since I’m not sure it is. Then, I know the pins/address for the Chitu stuff.

Is Bank4 the right setting for the FSMC? I would think this would change based on the address selection.

**Comment by jmz52** [comment]
@xC0000005 you are right, bank parameter depends on chip select pin.
MKS Robin is using NE4, so I've missed the hard-coded settings.

**Comment by xC0000005** [comment]
We're getting closer...
![IMG_1754](https://user-images.githubusercontent.com/32139633/84717685-6b7c8a80-af2b-11ea-9a6f-dcec43e2c53a.jpg)

**Comment by neonflux-io** [comment]
Hello, Nice to meet you

I am very interested in this job.
Please contact with me
